[benchmark] Janitor Duty: Datalore Legacy

palimondo opened the pull request:
This PR follows-up #20861 and #21413. It's a next batch of benchmark clean-up to enable [robust performance measurements](https://forums.swift.org/t/towards-robust-performance-measurement/11490) by adjusting workloads to run in reasonable time (< 1000 μs), minimizing the accumulated error. To maintain long-term performance tracking, it applies [legacy factor](https://github.com/apple/swift/pull/20212) where necessary.

`DataBenchmarks` have  been recently extended with a lot of variants in #20396, while some of the tests have regressed with regards to elimination of setup overhead (previously fixed in ae9f5f18b02dda11ce07099fb8a107ee43d64d36). Many `Large` variants (problematic as well as some perfectly fine ones) have been disabled in #20411.

Before application of `legacyFactor`, it was necessary to clean up the file to ease future maintenance.  I have applied the technique of shared test functions and inlined run functions pioneered by @lorentey in [SetTests](https://github.com/apple/swift/pull/18928/commits/6426e7e14cd8659da7b01fc6fb7be6ed31d6b82a#diff-9f29b15b063e8204f5440c09b6fdd940R36).

The disabled `Large` variants that were using `.veryLarge` `sampleData` were removed. We might re-introduce proper benchmarks for the `LargeSlice` implementation later, but for **that** the capacity has to be at least Int32.max (i.e. over 2GB). The `sampleData(.veryLarge)` was "only" 1GB. 

The disabled `Large` tests that were using `.large` `sampleData` were re-enabled to maintain the continuity of long-term performance tracking, since these benchmarks predated the addition of `Large` variants form #20396 (disabled in #20411).

`DataCreate[Small,Medium]` benchmarks were dominated by `arc4random_buf`. We don’t need truly random data, so an increasing sequence of bytes is used to fills the buffer instead. This changes their runtime, but I didn't rename them, as they were introduce very recently — I believe this doesn't disrupt any long-term performance tracking yet.

palimondo committed:
[Gardening] DataBenchmarks: quad to double space

palimondo committed:
[benchmark] Fix: Don’t benchmark arc4random_buf

arc4random_buf was dominating the creation time. We don’t need trully random data. Anything that quickly fills the buffer will do.

palimondo committed:
[Gardening] DataBenchmarks: format BenchmarkInfo

palimondo committed:
[Gardening] DataBenchmarks remove unused string

palimondo committed:
[benchmark] DataAppendArray refactored

Refactored to use shareable test method and inlined runFunction.

palimondo committed:
[benchmark] DataReset refactored

Refactored to use shareable test method and inlined runFunction.

palimondo committed:
[benchmark] DataReplace refactored

Refactored to use shared test method and inlined runFunctions.

phausler commented:
Those regressions seem like they are hitting the wrong code paths.

The other ones that improved are worrying too since they are really fast items and I’d wager the reduced time will cause more noise since the sample size isn’t a large enough significant sample.

Particularly these:
StringToDataSmall	700	50	-92.9%	14.00x
StringToDataEmpty	600	50	-91.7%	12.00x
StringToDataMedium	2900	250	-91.4%	11.60x
DataToStringEmpty	2300	200	-91.3%	11.50x
DataToStringSmall	4200	400	-90.5%	10.50x
DataToStringMedium	10800	1100	-89.8%	9

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo committed:
[benchmark] DataAppendSequence refactored

Refactored to use shareable test method and inlined runFunction.

palimondo committed:
[benchmark] DataToString, StringToData refactored

Refactored to use shared test method and inlined runFunctions.

palimondo committed:
[benchmark] Data workload variables

For extracting setup overhead.
Even the `large` sample Data is only 40KB, we can afford this to be static constant.

palimondo committed:
[benchmark] Remove incorrect `Large` variants

Removed the disabled Large variants (tagged `.skip`).

We might re-introduce proper benchmarks for the `LargeSlice` implementation later, but for **that** the capacity has to be at least Int32.max (i.e. over 2GB). The `sampleData(.veryLarge)` was only 1GB.

palimondo committed:
[benchmark] DataCount refactored

Refactored to use shared test method and inlined runFunctions.
Extracted setup overhead.

`blackHole` is ubiquitous in SBS, no need to state the obvious.

palimondo committed:
[benchmark] DataAccessBytes refactored

Refactored to use shared test method and inlined runFunctions.
Extracted setup overhead.

`blackHole` is ubiquitous in SBS, no need to state the obvious.

palimondo committed:
[benchmark] DataMutateBytes refactored

Refactored to use shared test method and inlined runFunctions.
Extracted setup overhead.

palimondo committed:
[benchmark] DataAppendBytes refactored

Refactored to use shared test method and inlined runFunctions.
Extracted setup overhead.

palimondo committed:
[benchmark] DataCreateArray refactored

Refactored to use inlined runFunctions.

palimondo committed:
[benchmark] DataCreate refactored

Refactored to use inlined runFunctions.

palimondo committed:
[benchmark] DataSetCount refactored

Refactored to use shared test method and inlined runFunctions.
Extracted setup overhead.

Removed `Large` variant, as it was testing the same underlying implementation as `Medium`.

palimondo committed:
[benchmark] DataCopyBytes refactored

Refactored to use shared test method and inlined runFunctions.
Extracted setup overhead.

Removed `Large` variant, as it was testing the same underlying implementation as `Medium`.

palimondo committed:
[benchmark] DataAppendData refactored

Refactored to use shared test method and inlined runFunctions.
Extracted setup overhead.

Re-enabled `Large` tests errorneously disabled in #20411.

palimondo committed:
[benchmark] DataSubscript refactored

Refactored to use inlined runFunctions.
Extracted setup overhead.

palimondo committed:
[benchmark] DataReplaceBuffer refactored

Refactored to use shared test method and inlined runFunctions.

Re-enabled `Large` test errorneously disabled in #20411. Removed `skip` tags, as this was the last use.

palimondo committed:
[benchmark] Move loop multipliers to BenchmarkInfo

palimondo committed:
[benchmark] DataBenchmarks Legacy Factor

Adjusted the multipliers to run in under 1000 μs.

The inner loop multipliers are divided by the `legacyFactor` to achieve the same reported runtime, but in a shorter time and with better precision (less accumulated error).

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo committed:
[benchmark] Fix DataSubscript fragile inlining?

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@phausler The String benchmarks were just an error on my part, I've lost a zero during loop multiplier extraction… 🤷‍♂️

I've finally realized what you meant by:
> Those regressions seem like they are hitting the wrong code paths.

Initially, I was puzzling over [`Data.swift`](https://github.com/apple/swift-corelibs-foundation/blob/master/Foundation/Data.swift) at `swift-corelibs-foundation`, not knowing there's a newer implementation right here in [`stdlib/public/Darwin/Foundation/Data.swift`](https://github.com/apple/swift/blob/master/stdlib/public/Darwin/Foundation/Data.swift), with a slew of inlinable implementation variants. Now I see that #20396 was adding benchmarks needed to validate improvements from #20225. It would have helped a lot if this link was mentioned in those PRs.

[You've mentioned before](https://github.com/apple/swift/pull/20396#issuecomment-449062485) that:
> Large variants are useful in the regards that it takes a completely different code path and has different performance characteristics

If I understand correctly, these were meant to exercise `LargeSlice` `_Representation`, right? But they did not — these were all [`InlineSlice`s](https://github.com/apple/swift/blob/9e9a791e94a6ee287445fb5eff4ef117e7a88c7f/stdlib/public/Darwin/Foundation/Data.swift#L898)! Therefore I have **removed the disabled `Large`** benchmark variants that were using `.veryLarge` `sampleData`. We might re-introduce proper benchmarks for the `LargeSlice` implementation later, but for that the capacity has to be at least `Int32.max` (i.e. over 2GB). The `sampleData(.veryLarge)` was "only" 1GB.

IMHO there are still multiple issues with benchmarking the initialization of Data, even after replacing `arc4random` which was dominating their runtime. My replacement works OK in optimized builds (it generates vectorized code) but fails quite badly in `Onone`. [You said that](https://github.com/apple/swift/pull/20396#issuecomment-449062485):

> The arc4random is being used to ensure the kernel does not just hand out a non faulting page of memory. Just being an allocation is not enough to properly identify used cases. It must be written to: and patterns like repeated are a completely different code path in the sequence initialization.

I have played with many different variants: `Data.init(capacity:)` and fully filling it with byte sequence, or `Data.init(count:)` and just writing 1 byte per page. It seems to take about the same time. But this is all utterly impractical for testing `LargeSlice`. I think we should instead rely on the page-faulted constant zero initialized `Data.init(count:)` which is still OK, when we want to compare relative costs between the various `Data` `_Representation`s. What do you think?

It also looks like the "inline everything" strategy is backfiring on `DataSubscript` benchmarks, as @airspeedswift warned @itaiferber [during review for `swift-5.0-branch`](https://github.com/apple/swift/pull/21292#issuecomment-447396273). Compare the results 
1df944c did above and below here.

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@eeckstein @phausler Please review 🙏.
The step-by-step refactoring is best reviewed by [individual commits](https://github.com/apple/swift/pull/21516/commits/33873284212e3530782fa785f1323e1d6acc70a6).

eeckstein reviewed (approved):
lgtm, but please wait for @phausler to review.

itaiferber reviewed (commented):
Changes here look good to me too, though I'd want @phausler to get a chance to review. Tiny nits inline; otherwise, do you mind commenting on how you decided on these legacy factors?

I'm also separately working on pulling some of the inlining out where it was excessive, so I'll try to test with the changes here so we're looking at a consistent view of the tests.

phausler commented on benchmark/single-source/DataBenchmarks.swift:
```diff
@@ -13,638 +13,345 @@
+let d: [BenchmarkCategory] =  [.validation, .api, .Data]
+
+/// Workload scaling: Inner loop multipliers
+let M = 10_000  /// default
```
The only quibble I have here is that these could use a bit better set of names.

palimondo replied:
We should probably "precompute and inline" them all. I had extracted these constants only to demonstrate how's the `legacyFactor` derived. 

Would you be OK with having the multipliers all inline, or do you have a suggestion for better names for the constants?

palimondo replied:
Going with inlined constants, unless you wan't me to revert back to constants.

itaiferber commented on benchmark/single-source/DataBenchmarks.swift:
```diff
@@ -56,72 +58,90 @@ public let DataBenchmarks = [
+    legacyFactor: 20),
   BenchmarkInfo(name: "DataAppendDataMediumToLarge",
-    runFunction: { append($0*M, data: medium, to: large) }, tags: d),
+    runFunction: { append($0*50, data: medium, to: large) }, tags: d),
```
Oh, I see that these are no longer dependent on `M` at all, and are thus not adjusted by the legacy factor?

palimondo replied:
I didn't add the factor here, because it was `.skip`ped before my refactoring. But since it was historically in the "database" before it was accidentally(?) disabled with other `Large` tests, I should probably add the `legacyFactor` anyway.

palimondo replied:
@eeckstein ^^ Thoughts?

palimondo replied:
I'm adding LF for these two cases too. Can't hurt ;-)

phausler reviewed (approved):
The scalars could use a bit of renaming, but functionally this looks reasonable to me.

palimondo committed:
[benchmark] Inlined loop miltipliers

Also added 2 forgotten legacy factors.

palimondo commented:
@phausler, @itaiferber Could you also respond to my suggestion about testing `LargeSlice` and `init` [from above](https://github.com/apple/swift/pull/21516#issuecomment-451601361)?

palimondo commented:
Here are few `init` benchmark variations I've been playing with. These are using new naming convention from #20334. Note that the loop multiplier differs between the groups.
<details>
<summary><code>Data.init.[count,capacity]</code> variants</summary>

````swift
… 
  BenchmarkInfo(name: "Data.init.Empty",
    runFunction: { init_($0*10_000, size: .empty) }, tags: d),
  BenchmarkInfo(name: "Data.init.Small",
    runFunction: { init_($0*100, size: .small) }, tags: d),
  BenchmarkInfo(name: "Data.init.Medium",
    runFunction: { init_($0*100, size: .medium) }, tags: d),

  BenchmarkInfo(name: "Data.init.count.Filled.Empty",
    runFunction: { init_($0*100, count: 0) }, tags: d),
  BenchmarkInfo(name: "Data.init.count.Filled.Small",
    runFunction: { init_($0*100, count: 11) }, tags: d),
  BenchmarkInfo(name: "Data.init.count.Filled.Medium",
    runFunction: { init_($0*100, count: 1033) }, tags: d),

  BenchmarkInfo(name: "Data.init.capacity.Filled.Empty",
    runFunction: { init_($0*100, capacity: 0) }, tags: d),
  BenchmarkInfo(name: "Data.init.capacity.Filled.Small",
    runFunction: { init_($0*100, capacity: 11) }, tags: d),
  BenchmarkInfo(name: "Data.init.capacity.Filled.Medium",
    runFunction: { init_($0*100, capacity: 1033) }, tags: d),

  // init(count: Int)
  // Creates a new data buffer with the specified count of zeroed bytes.
  BenchmarkInfo(name: "Data.init.count.Inlined.Empty",
    runFunction: { for _ in 0..<$0*10_000 { blackHole(Data(count: 0)) } },
    tags: d),
  BenchmarkInfo(name: "Data.init.count.Inlined.Small",
    runFunction: { for _ in 0..<$0*10_000 { blackHole(Data(count: 11)) } },
    tags: d),
  BenchmarkInfo(name: "Data.init.count.Inlined.Medium",
    runFunction: { for _ in 0..<$0*100 { blackHole(Data(count: 1033)) } },
    tags: d),

  // init(capacity: Int)
  // Creates an empty data buffer of a specified size.
  BenchmarkInfo(name: "Data.init.capacity.Inlined.Small",
    runFunction: { for _ in 0..<$0*100 {
      let size = 11
      var data = Data(capacity: size)
      data.withUnsafeMutableBytes { (bytes: UnsafeMutablePointer<UInt8>) -> () in
        for i in 0..<size {
          bytes[i] = UInt8(truncatingIfNeeded: i)
        }
      }
      blackHole(d)
    } },
    tags: d),
  BenchmarkInfo(name: "Data.init.capacity.Inlined.Medium",
    runFunction: { for _ in 0..<$0*100 {
      let size = 1033
      var data = Data(capacity: size)
      data.withUnsafeMutableBytes { (bytes: UnsafeMutablePointer<UInt8>) -> () in
        for i in 0..<size {
          bytes[i] = UInt8(truncatingIfNeeded: i)
        }
      }
      blackHole(d)
    } },
    tags: d),

…

@inline(never)
public func init_(_ N: Int, count: Int) {
  for _ in 1...N {
    var data = Data(count: count)
    data.withUnsafeMutableBytes { (bytes: UnsafeMutablePointer<UInt8>) -> () in
      for i in 0..<count {
        bytes[i] = UInt8(truncatingIfNeeded: i)
      }
    }
    blackHole(data)
  }
}

@inline(never)
public func init_(_ N: Int, capacity: Int) {
  for _ in 1...N {
    var data = Data(capacity: capacity)
    data.withUnsafeMutableBytes { (bytes: UnsafeMutablePointer<UInt8>) -> () in
      for i in 0..<capacity {
        bytes[i] = UInt8(truncatingIfNeeded: i)
      }
    }
    blackHole(data)
  }
}
````

</details>

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

itaiferber commented:
@palimondo Regarding the large slice cases — I personally believe that both large slices and large _inline_ slices are worth testing, but perhaps not at the cost of destabilizing other tests, or causing unnecessary delays in benchmarking for everyone. With more extensive, separate benchmarking infrastructure (to make it more reasonable to test this case more sparingly), it might be worth investigating. I think we're okay to drop the tests for now though.

palimondo commented:
Perhaps I wasn't being clear enough. My point was, that I believe the point @phausler was making about the `.veryLarge` test wasn't really necessary:

> The arc4random is being used to ensure the kernel does not just hand out a non faulting page of memory. Just being an allocation is not enough to properly identify used cases. 

When I played with `Data.init(count:)` I saw how kernel cheats and gives you a lot of memory in almost no time. The huge delay comes first when writing to it. So I think we can use this to our advantage: quickly get the `LargeSlice` representation of `Data` backed with faulting pages. But that doesn't really matter for the purposes of our tests, because we don't want to be benchmarking  how quickly kernel copies memory. We are interested in the relative performance of the underlying `_Representations`s and even though they might be backed by virtual memory subsystem's trickery, as long our workloads are the same (expanding the count by same amount or replacing the same amount of bytes, etc) we are correctly comparing the relative costs of `InlineSlice` vs. `LargeSlice`. Am I mistaken in my assumption here?

phausler commented:
The perf differential tween `InlineSlice` and `LargeSlice` has to do with the indirection to the additional ref type storing the range. Ideally we want the performance of that indirection to be similar to each other, and at the very least to be dwarfed by normal access. Unfortunately the benchmarks as they are don't give us that type of granularity and are likely useful to be done in an instrumentation mechanism. Most of these benchmarks are suitable to be stuck in an infinite (or practically infinite) loop  and be profiled via tools like Instruments.
